### PR TITLE
[SPARK-35886][SQL] CodeGenerator.getLocalInputVariableValues should handle case that expression match  SubExprEliminationState but not  VariableValue

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
For test case 
```
spark.sql(
  """
    |CREATE TABLE t1 (
    |  c1 DECIMAL(18,6),
    |  c2 DECIMAL(18,6),
    |  c3 DECIMAL(18,6))
    |USING parquet;
    |""".stripMargin)

spark.sql("SELECT sum(c1 * c3) + sum(c2 * c3) FROM t1").show
```
failed when codegen
```
20:23:36.272 ERROR org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 56, Column 6: Expression "agg_exprIsNull_2_0" is not an rvalue
org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 56, Column 6: Expression "agg_exprIsNull_2_0" is not an rvalue
	at org.codehaus.janino.UnitCompiler.compileError(UnitCompiler.java:12675)
	at org.codehaus.janino.UnitCompiler.toRvalueOrCompileException(UnitCompiler.java:7676)

```

It's caused by when call CodeGenerator.getLocalInputVariableValues, expression match `SubExprEliminationState` but is not  `VariableValue`,  then skip this whole expression, but when call `splitAggregateExpressions`,  since we need to find all input variables by this method, this will cause we skip some inputs.
Here in this case , we put expression's children to stack too.

### Why are the changes needed?
Fix Bug

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan @maropu @wangyum @Ngone51

**Comment by wangyum** [comment]
Also cc @viirya

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -4042,6 +4042,22 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
         Row(1, 2, 1, 2) :: Nil)
     }
   }
+
+  test("SPARK-35886: CodeGenerator.getLocalInputVariableValues should handle " +
+    "matched subQuery but not VariableValue") {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4047]: you mean but not ONLY `VariableValue`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 2]
```diff
@@ -1766,19 +1766,20 @@ object CodeGenerator extends Logging {
...
 
-    val stack = mutable.Stack[Expression](expr)
     while (stack.nonEmpty) {
       stack.pop() match {
         case e if subExprs.contains(e) =>
           val SubExprEliminationState(isNull, value) = subExprs(e)
-          collectLocalVariable(value)
-          collectLocalVariable(isNull)
+          collectLocalVariable(value, e)
+          collectLocalVariable(isNull, e)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:1782]: Oh, some `SubExprEliminationState` contains not `VariableValue` but still some statements that includes other expressions?

**Comment by viirya** [comment]
Do we have "subquery" here? Please update the description if it is not "subquery".

**Review by viirya — COMMENTED** [review]
Looks reasonable, although the description and the title confuse.

**Comment by AngersZhuuuu** [comment]
> Do we have "subquery" here? Please update the description if it is not "subquery".

My mistake, should be  SubExprEliminationState

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:1782]: > Oh, some `SubExprEliminationState` contains not `VariableValue` but still some statements that includes other expressions?

Yes, the expression contains input var  may match `SubExprEliminationState` and just skiped.

**Comment by AngersZhuuuu** [comment]
> Looks reasonable, although the description and the title confuse.

Updated, how about current?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:1782]: Hmm, after second look, it looks not root cause. A subexpr should be evaluated before using it. 

So technically all its children expressions are not needed.

I found the root cause. `PromotePrecision` overwrites `Expression.genCode` where is subexpression replace happens. So `PromotePrecision` skips subexpression elimination. I will submit another PR and include your test there. Thanks.

**Comment by AngersZhuuuu** [comment]
Close this according to  https://github.com/apache/spark/pull/33103

**Pull request closed without merging**